// raft-js/src/instructions/disable-stepper-motor.ts
/**
 * @module RaftInstruction
 */ /** */

import RaftInstruction, {RaftInstructionParams} from "./base";

export interface DisableStepperMotorInstructionParams extends RaftInstructionParams {
    x?: boolean,
    y?: boolean,
    z?: boolean,
    a?: boolean,
    b?: boolean,
    e?: boolean,
}

/**
 * Disable Stepper Motors
 * <p>Disable some or all of the stepper motors, allowing the respective axes
 * to move "freely".</p>
 */
export default class DisableStepperMotorInstruction extends RaftInstruction {

    readonly x: boolean;
    readonly y: boolean;
    readonly z: boolean;
    readonly a: boolean;
    readonly b: boolean;
    readonly e: boolean;

    constructor (params: DisableStepperMotorInstructionParams) {
        super(params);
        if (params.x === undefined && params.y === undefined && params.z === undefined
            && params.a === undefined && params.b === undefined && params.e === undefined) {
            this.x = true;
            this.y = true;
            this.z = true;
            this.a = true;
            this.b = true;
            this.e = true;
        } else {
            this.x = (params.x === undefined ? false : !!params.x);
            this.y = (params.y === undefined ? false : !!params.y);
            this.z = (params.z === undefined ? false : !!params.z);
            this.a = (params.a === undefined ? false : !!params.a);
            this.b = (params.b === undefined ? false : !!params.b);
            this.e = (params.e === undefined ? false : !!params.e);
        }
    }

    toString(): string {
        return "DisableStepperMotorInstruction";
    }

}
