// raft-js/src/instructions/home.ts
/**
 * @module RaftInstruction
 */ /** */

import MoveInstruction, {MoveInstructionParams} from "./move";

export interface HomeInstructionParams extends MoveInstructionParams {
    x?: boolean | number,
    y?: boolean | number,
    z?: boolean | number,
    feedrate?: number,
    toMaximum?: boolean
}

/**
 * Move to Origin (Home)
 * <p>Move all (or only the supplied) axes back to the zero endstops, at maximum speed
 * (or a specified feedrate) and towards the axis minimum by default (unless specified).</p>
 */
export default class HomeInstruction extends MoveInstruction {

    readonly x: boolean | number;
    readonly y: boolean | number;
    readonly z: boolean | number;
    readonly toMaximum: boolean;

    constructor (params: HomeInstructionParams) {
        super(params);
        if (params.x === undefined && params.y === undefined && params.z === undefined) {
            this.x = true;
            this.y = true;
            this.z = true;
        } else {
            this.x = params.x;
            this.y = params.y;
            this.z = params.z;
        }
        this.toMaximum = (params.toMaximum === undefined ? false : params.toMaximum);
    }

    toString(): string {
        return "HomeInstruction";
    }

}
