// raft-js/src/instructions/load-axis-offset.ts
/**
 * @module RaftInstruction
 */ /** */

import InvalidArgumentError from "../errors/invalid-argument";
import RaftInstruction, {RaftInstructionParams} from "./base";
import {validateNumber} from "../common/validation";

export interface LoadAxisOffsetInstructionParams extends RaftInstructionParams {
    x?: boolean,
    y?: boolean,
    z?: boolean,
    a?: boolean,
    b?: boolean
}

export default class LoadAxisOffsetInstruction extends RaftInstruction {

    readonly x: boolean;
    readonly y: boolean;
    readonly z: boolean;
    readonly a: boolean;
    readonly b: boolean;

    constructor (params: LoadAxisOffsetInstructionParams) {
        super(params);
        if (params.x === undefined && params.y === undefined && params.z === undefined
            && params.a === undefined && params.b === undefined) {
            this.x = true;
            this.y = true;
            this.z = true;
            this.a = true;
            this.b = true;
        } else {
            this.x = !!params.x;
            this.y = !!params.y;
            this.z = !!params.z;
            this.a = !!params.a;
            this.b = !!params.b;
        }
    }

    toString(): string {
        return "LoadAxisOffsetInstruction";
    }

}
