// raft-js/src/instructions/o-materials-used.ts
/**
 * @module RaftInstruction
 */ /** */

import OEMInstruction, {OEMInstructionParams} from "./oem";

export interface MaterialParams {
    index: number
    color?: [number, number, number]
    name?: string
}

export class Material {

    readonly index: number;
    readonly color: [number, number, number];
    readonly name: string;

    constructor (params: MaterialParams) {
        if (params.index === 0) {
            this.index = params.index;
            this.color = [0, 0, 0];
            this.name = "";
        } else {
            this.index = params.index;
            this.color = params.color;
            this.name = params.name;
        }
    }

}

export interface OMaterialsUsedInstructionParams extends OEMInstructionParams {
    materials: [Material, Material, Material, Material]
}

export default class OMaterialsUsedInstruction extends OEMInstruction {

    readonly materials: [Material, Material, Material, Material];

    constructor (params: OMaterialsUsedInstructionParams) {
        super(params);
        this.materials = params.materials;
    }

    toString(): string {
        return "OMaterialsUsedInstruction";
    }

}
