// raft-js/src/instructions/set-axis-travel-bound.ts
/**
 * @module RaftInstruction
 */ /** */

import InvalidArgumentError from "../errors/invalid-argument";
import RaftInstruction, {RaftInstructionParams} from "./base";
import {validateNumber} from "../common/validation";

export interface SetAxisTravelBoundInstructionParams extends RaftInstructionParams {
    isMaximum?: boolean,
    x?: number,
    y?: number,
    z?: number
}

export default class SetAxisTravelBoundInstruction extends RaftInstruction {

    readonly isMaximum: boolean;
    readonly x: number;
    readonly y: number;
    readonly z: number;

    constructor (params: SetAxisTravelBoundInstructionParams) {
        if (params.x !== undefined && !validateNumber(params.x)) {
            throw new InvalidArgumentError("Invalid x value '" + params.x + "' supplied as argument");
        }
        if (params.y !== undefined && !validateNumber(params.y)) {
            throw new InvalidArgumentError("Invalid y value '" + params.y + "' supplied as argument");
        }
        if (params.z !== undefined && !validateNumber(params.z)) {
            throw new InvalidArgumentError("Invalid z value '" + params.z + "' supplied as argument");
        }
        super(params);
        this.isMaximum = (params.isMaximum === undefined ? true : !!params.isMaximum);
        this.x = (params.x === undefined ? null : params.x);
        this.y = (params.y === undefined ? null : params.y);
        this.z = (params.z === undefined ? null : params.z);
    }

    toString(): string {
        return "SetAxisTravelBoundInstruction";
    }

}
