// raft-js/lib/engines/ultigcode.js
"use strict";
/**
 * @module RaftEngine
 */ /** */
Object.defineProperty(exports, "__esModule", { value: true });
const marlin_1 = require("./marlin");
class UltiGCodeEngine extends marlin_1.default {
    constructor(options) {
        options.volumetricExtrusion = true;
        super(options);
    }
    clone() {
        let copy = new UltiGCodeEngine({
            toolheadCount: this.toolheadCount,
            independentExtruders: this.independentExtruders,
            volumetricExtrusion: this.volumetricExtrusion,
            defaultHomeDirectionMax: this.defaultHomeDirectionMax,
            filamentDiameter: this.filamentDiameter,
            initialXYZCoordinateMode: this.stateMachines["x"].getCurrentState().coordinateMode,
            initialExtrusionCoordinateMode: this.stateMachines["extrusion"].getCurrentState().coordinateMode
        });
        this.copyStateMachinesTo(copy);
        return copy;
    }
    toRaftInstruction(line) {
        return super.toRaftInstruction(line);
    }
    toOutputInstruction(instruction) {
        return super.toOutputInstruction(instruction);
    }
    toString() {
        return "UltiGCodeEngine";
    }
}
exports.default = UltiGCodeEngine;
