// raft-js/lib/enums.js
"use strict";
/**
 * @module Raft
 */ /** */
Object.defineProperty(exports, "__esModule", { value: true });
var CoordinateModes;
(function (CoordinateModes) {
    CoordinateModes[CoordinateModes["None"] = 0] = "None";
    CoordinateModes[CoordinateModes["Absolute"] = 1] = "Absolute";
    CoordinateModes[CoordinateModes["Relative"] = 2] = "Relative";
})(CoordinateModes = exports.CoordinateModes || (exports.CoordinateModes = {}));
var DistanceUnits;
(function (DistanceUnits) {
    DistanceUnits[DistanceUnits["Millimeters"] = 0] = "Millimeters";
    DistanceUnits[DistanceUnits["Inches"] = 1] = "Inches";
})(DistanceUnits = exports.DistanceUnits || (exports.DistanceUnits = {}));
var SpeedUnits;
(function (SpeedUnits) {
    SpeedUnits[SpeedUnits["MillimetersPerMinute"] = 0] = "MillimetersPerMinute";
    SpeedUnits[SpeedUnits["InchesPerMinute"] = 1] = "InchesPerMinute";
})(SpeedUnits = exports.SpeedUnits || (exports.SpeedUnits = {}));
var PrintStates;
(function (PrintStates) {
    PrintStates[PrintStates["None"] = 0] = "None";
    PrintStates[PrintStates["StartSequence"] = 1] = "StartSequence";
    PrintStates[PrintStates["EndSequence"] = 2] = "EndSequence";
    PrintStates[PrintStates["Skirt"] = 3] = "Skirt";
    PrintStates[PrintStates["Raft"] = 4] = "Raft";
    PrintStates[PrintStates["Brim"] = 5] = "Brim";
    PrintStates[PrintStates["Support"] = 6] = "Support";
    PrintStates[PrintStates["InnerPerimeter"] = 7] = "InnerPerimeter";
    PrintStates[PrintStates["OuterPerimeter"] = 8] = "OuterPerimeter";
    PrintStates[PrintStates["SolidLayer"] = 9] = "SolidLayer";
    PrintStates[PrintStates["Infill"] = 10] = "Infill";
    PrintStates[PrintStates["GapFill"] = 11] = "GapFill";
})(PrintStates = exports.PrintStates || (exports.PrintStates = {}));
