// raft-js/lib/instructions/enable-stepper-motor.js
"use strict";
/**
 * @module RaftInstruction
 */ /** */
Object.defineProperty(exports, "__esModule", { value: true });
const base_1 = require("./base");
/**
 * Enable Stepper Motors
 * <p>Enable some or all of the stepper motors.</p>
 */
class EnableStepperMotorInstruction extends base_1.default {
    constructor(params) {
        super(params);
        if (params.x === undefined && params.y === undefined && params.z === undefined
            && params.a === undefined && params.b === undefined && params.e === undefined) {
            this.x = true;
            this.y = true;
            this.z = true;
            this.a = true;
            this.b = true;
            this.e = true;
        }
        else {
            this.x = (params.x === undefined ? false : !!params.x);
            this.y = (params.y === undefined ? false : !!params.y);
            this.z = (params.z === undefined ? false : !!params.z);
            this.a = (params.a === undefined ? false : !!params.a);
            this.b = (params.b === undefined ? false : !!params.b);
            this.e = (params.e === undefined ? false : !!params.e);
        }
    }
    toString() {
        return "EnableStepperMotorInstruction";
    }
}
exports.default = EnableStepperMotorInstruction;
