// raft-js/lib/instructions/set-temperature.js
"use strict";
/**
 * @module RaftInstruction
 */ /** */
Object.defineProperty(exports, "__esModule", { value: true });
const invalid_argument_1 = require("../errors/invalid-argument");
const base_1 = require("./base");
const validation_1 = require("../common/validation");
class SetTemperatureInstruction extends base_1.default {
    constructor(params) {
        if (params.temperature === undefined && params.maxTargetTemperature === undefined) {
            throw new invalid_argument_1.default("Temperature argument missing from instruction");
        }
        if (params.temperature !== null && params.temperature !== undefined && (!validation_1.validateNumber(params.temperature) || params.temperature < 0)) {
            throw new invalid_argument_1.default("Invalid temperature '" + params.temperature + "' supplied as argument");
        }
        if (params.maxTargetTemperature !== undefined && (!validation_1.validateNumber(params.maxTargetTemperature) || params.maxTargetTemperature < 0)) {
            throw new invalid_argument_1.default("Invalid target temperature '" + params.maxTargetTemperature + "' supplied as argument");
        }
        if (params.stabilizeTimeout !== undefined && (!validation_1.validateNumber(params.stabilizeTimeout) || params.stabilizeTimeout < 0)) {
            throw new invalid_argument_1.default("Invalid stabilize timeout '" + params.stabilizeTimeout + "' supplied as argument");
        }
        super(params);
        this.temperature = (params.temperature === undefined ? null : params.temperature);
        this.maxTargetTemperature = (params.maxTargetTemperature === undefined ? null : params.maxTargetTemperature);
        this.stabilize = (params.stabilize === undefined ? true : !!params.stabilize);
        if (this.stabilize) {
            this.stabilizeTimeout = (params.stabilizeTimeout === undefined ? null : params.stabilizeTimeout);
        }
        else {
            this.stabilizeTimeout = null;
        }
    }
    toString() {
        return "SetTemperatureInstruction";
    }
}
exports.default = SetTemperatureInstruction;
